src/dcegm/egm.py

```diff
@@ -0,0 +1,458 @@
+            Position [.., 0, :] contains the endogenous grid over wealth M,
+            and [.., 1, :] stores the corresponding value of the policy function
+            c(M, d), for each state and each discrete choice.
+        value_array (np.ndarray): Multi-dimensional np.ndarray storing the
+            choice-specific value functions; of shape
+            [n_states, n_discrete_choices, 2, 1.1 * (n_grid_wealth + 1)].
+            Position [.., 0, :] contains the endogenous grid over wealth M,
+            and [.., 1, :] stores the corresponding value of the value function
+            v(M, d), for each state and each discrete choice.
+        compute_next_wealth_matrices (callable): User-defined function to compute the
+            agent's wealth matrices of the next period (t + 1). The inputs
+            ```savings_grid```, ```income_shocks```, ```params``` and ```options```
+            are already partialled in.
+        compute_marginal_utility (callable): User-defined function to compute the
+            agent's marginal utility. The input ```params``` is already partialled in.
+        compute_value (callable): User-defined function to compute
+            the agent's value function in the credit-constrained area. The inputs
+            ```params``` and ```compute_utility``` are already partialled in.
+        get_state_specific_choice_set (Callable): User-supplied function returning for
+            each state all possible choices.
+
+    Returns:
+
+    """
+    child_state_index = state_indexer[tuple(child_state)]
+    choice_policies_child = policy_array[child_state_index]
+    choice_values_child = value_array[child_state_index]
+
+    child_node_choice_set = get_state_specific_choice_set(
+        child_state, state_space, state_indexer
+    )
+    next_period_wealth = compute_next_wealth_matrices(
+        child_state,
+        savings_grid=exogenous_savings_grid,
+        income_shock=income_shock_draws,
+    )
+    # Interpolate next period policy and values to match the
+    # contemporary matrix of potential next period wealths
+    child_policy = get_child_state_choice_specific_policy(
+        child_node_choice_set,
+        next_period_wealth,
+        next_period_policy=choice_policies_child,
+    )
+    choice_child_values = get_child_state_choice_specific_values(
+        child_node_choice_set,
+        next_period_wealth=next_period_wealth,
+        next_period_value=choice_values_child,
+        compute_value=compute_value,
+    )
+
+    child_state_marginal_utility = get_child_state_policy(
+        child_node_choice_set,
+        next_period_policy=child_policy,
+        next_period_value=choice_child_values,
+        taste_shock_scale=taste_shock_scale,
+        compute_marginal_utility=compute_marginal_utility,
+    )
+
+    child_state_log_sum = calc_exp_max_value(choice_child_values, taste_shock_scale)
+
+    marginal_utility_weighted = (
```

Comment:
I did some refactoring and named the outputs. Are these good names? They are not what I expect given the function name ```get_child_state_policy_and_value```. We should change the function name or rename the variables.

Reply:
Renamed variables and function. This will change once we scalarize, so good for now I would say!